docs/adr/solana-storage-architecture.md

```diff
@@ -0,0 +1,475 @@
+
+**Merkle Root for Archival**:
+
+- During operation: Store bitmap + individual commitments
+- At window slide: Compute merkle root from stored commitments
+- Archive merkle root for efficient batch verification of historical data
+- Individual commitments provide real-time verification capability
+
+This approach provides:
+
+- Immediate packet verification via stored commitments
+- Efficient historical verification via merkle roots
+- No expensive merkle tree updates during packet sends
+
+**Implementation**:
+
+```rust
+pub struct RouterState {
+    // Primary window
+    pub sent_packets: BitVec,         // 1 = active (sent, not finalized), 0 = not sent or finalized
+    pub window_start: u64,            // First sequence in window
+    pub window_size: u64,             // Size of sliding window
+
+    // Packet commitments for verification (32 bytes per packet)
+    pub packet_commitments: BTreeMap<u64, [u8; 32]>, // sequence -> hash(packet)
+
+
+    // Active packets outside current window (sparse storage)
+    pub active_packets: Vec<u64>, // Just sequence numbers of active packets
+    pub active_packet_commitments: BTreeMap<u64, [u8; 32]>,
+}
+
+// Packet commitment storage - separate PDAs for each commitment
+// PDA: ["packet_commitment", channel_id.as_bytes(), sequence.to_le_bytes()]
+pub struct PacketCommitment {
+   pub commitment: [u8; 32],         // hash(packet)
+   pub timestamp: i64,               // For TTL enforcement
+}
+```
+
+**State Transitions**:
+
+- Send packet: Set bit to 1, create PacketCommitment PDA with hash
+- Receive acknowledgment: Set bit to 0, close PacketCommitment PDA to reclaim rent
+- Timeout packet: Verify commitment PDA exists and matches, set bit to 0, close PDA
+- Check if sent: Read bit (1 = yes and active, 0 = no or already finalized)
+- Window slide: Extract active packets to active_packets vector, close PDAs for finalized packets
+
+**Window Management**:
+
+- Sequence number maps to bitmap index: `index = sequence - window_start`
+- Window slides when ~80% full or when early portion has no active packets
+- Active packets are extracted to `active_packets` set before sliding
+- Only stores sequence numbers of in-flight packets outside window
+
+**Storage Bounds and Limits (example)**:
+
+- Window size: 10,000 packets
+- Maximum active packets outside window: 1,000
+- Packet TTL: 24 hours
+- Total commitment limit: 11,000 (window + active)
+
+Enforcement:
+
+- Reject sends when active packets exceed limit
+- Expire packets older than TTL
+- Prevent commitment map overflow
+
+This ensures:
+
+- Maximum 11,000 packets tracked at any time
+- Packets expire after 24 hours
+- Prevents DoS via unbounded storage growth
+
+**Storage Size Calculation**:
+
+```
+Base RouterState: ~200 bytes
+Bitmap (10,000 bits): 1,250 bytes
+Commitments (11,000 × 32): 352,000 bytes
+Active packets set (1,000 × 8): 8,000 bytes
+Total: ~361KB (well within 10MB limit)
+
+```
+
+### 3. Off-chain Archival System
```

Comment:
I won't comment much on the archival system at the moment since I have many comments on the on-chain state